File: 02_VBNET_Level1/練習問題/問題1_在庫管理バグ/InventoryManager/FormMain.vb
Imports System.Windows.Forms

Namespace InventoryManager
    Public Partial Class FormMain
        Inherits Form

        Private products As New List(Of Product)()
        Private nextId As Integer = 1

        Public Sub New()
            InitializeComponent()
        End Sub

        ' フォーム読み込み時にサンプルデータを追加
        Private Sub FormMain_Load(sender As Object, e As EventArgs) Handles MyBase.Load
            ' カテゴリのドロップダウン設定
            cmbCategory.Items.AddRange(New String() {"食品", "飲料", "日用品"})
            cmbCategory.SelectedIndex = 0

            ' サンプルデータ
            products.Add(New Product(nextId, "コーヒー豆", "食品", 50, 800))
            nextId += 1
            products.Add(New Product(nextId, "緑茶ペットボトル", "飲料", 3, 150))
            nextId += 1
            products.Add(New Product(nextId, "食器用洗剤", "日用品", 20, 350))
            nextId += 1
            products.Add(New Product(nextId, "カップラーメン", "食品", 100, 200))
            nextId += 1
            products.Add(New Product(nextId, "ミネラルウォーター", "飲料", 5, 100))
            nextId += 1

            RefreshGrid(products)
        End Sub

        ' 追加ボタン
        Private Sub btnAdd_Click(sender As Object, e As EventArgs) Handles btnAdd.Click
            ' バリデーション
            If String.IsNullOrWhiteSpace(txtProductName.Text) Then
                MessageBox.Show("商品名を入力してください", "入力エラー", MessageBoxButtons.OK, MessageBoxIcon.Warning)
                Return
            End If

            ' 在庫数チェック
            Dim stock As Integer
            If Not Integer.TryParse(txtStock.Text, stock) OrElse stock < 0 Then
                MessageBox.Show("在庫数は0以上の整数で入力してください", "入力エラー", MessageBoxButtons.OK, MessageBoxIcon.Warning)
                Return
            End If

            ' 単価チェック
            ' ★★★ バグ3: 単価のチェックが >= 0 になっている ★★★
            ' 仕様 FR-007: 単価は1以上の整数
            ' 実装: 0以上のチェックになっており、単価0円が登録できてしまう
            Dim price As Integer
            If Not Integer.TryParse(txtPrice.Text, price) OrElse price < 0 Then
                MessageBox.Show("単価は0以上の整数で入力してください", "入力エラー", MessageBoxButtons.OK, MessageBoxIcon.Warning)
                Return
            End If

            Dim product As New Product(nextId, txtProductName.Text, cmbCategory.SelectedItem.ToString(), stock, price)
            nextId += 1
            products.Add(product)

            ' 入力欄クリア
            txtProductName.Text = ""
            cmbCategory.SelectedIndex = 0
            txtStock.Text = ""
            txtPrice.Text = ""

            RefreshGrid(products)
        End Sub

        ' 編集ボタン
        Private Sub btnEdit_Click(sender As Object, e As EventArgs) Handles btnEdit.Click
            If dgvInventory.SelectedRows.Count = 0 Then
                MessageBox.Show("商品を選択してください", "警告", MessageBoxButtons.OK, MessageBoxIcon.Warning)
                Return
            End If

            ' バリデーション
            If String.IsNullOrWhiteSpace(txtProductName.Text) Then
                MessageBox.Show("商品名を入力してください", "入力エラー", MessageBoxButtons.OK, MessageBoxIcon.Warning)
                Return
            End If

            Dim stock As Integer
            If Not Integer.TryParse(txtStock.Text, stock) OrElse stock < 0 Then
                MessageBox.Show("在庫数は0以上の整数で入力してください", "入力エラー", MessageBoxButtons.OK, MessageBoxIcon.Warning)
                Return
            End If

            Dim price As Integer
            If Not Integer.TryParse(txtPrice.Text, price) OrElse price < 0 Then
                MessageBox.Show("単価は0以上の整数で入力してください", "入力エラー", MessageBoxButtons.OK, MessageBoxIcon.Warning)
                Return
            End If

            Dim selectedId As Integer = CInt(dgvInventory.SelectedRows(0).Cells("Id").Value)
            Dim product As Product = products.Find(Function(p) p.Id = selectedId)

            If product IsNot Nothing Then
                product.ProductName = txtProductName.Text
                product.Category = cmbCategory.SelectedItem.ToString()
                product.Stock = stock
                product.UnitPrice = price

                RefreshGrid(products)
            End If
        End Sub

        ' 削除ボタン
        Private Sub btnDelete_Click(sender As Object, e As EventArgs) Handles btnDelete.Click
            If dgvInventory.SelectedRows.Count = 0 Then
                MessageBox.Show("商品を選択してください", "警告", MessageBoxButtons.OK, MessageBoxIcon.Warning)
                Return
            End If

            Dim selectedId As Integer = CInt(dgvInventory.SelectedRows(0).Cells("Id").Value)
            Dim product As Product = products.Find(Function(p) p.Id = selectedId)

            If product IsNot Nothing Then
                Dim result As DialogResult = MessageBox.Show(
                    product.ProductName & "を削除しますか？",
                    "削除確認",
                    MessageBoxButtons.YesNo,
                    MessageBoxIcon.Question)

                If result = DialogResult.Yes Then
                    products.Remove(product)
                    RefreshGrid(products)
                End If
            End If
        End Sub

        ' 検索ボタン
        Private Sub btnSearch_Click(sender As Object, e As EventArgs) Handles btnSearch.Click
            Dim keyword As String = txtSearch.Text

            If String.IsNullOrWhiteSpace(keyword) Then
                RefreshGrid(products)
                Return
            End If

            Dim results As List(Of Product) = products.FindAll(
                Function(p) p.ProductName.IndexOf(keyword, StringComparison.OrdinalIgnoreCase) >= 0)

            If results.Count = 0 Then
                MessageBox.Show("該当する商品はありません", "検索結果", MessageBoxButtons.OK, MessageBoxIcon.Information)
            End If

            RefreshGrid(results)
        End Sub

        ' DataGridView選択変更時に入力欄に値をセット
        Private Sub dgvInventory_SelectionChanged(sender As Object, e As EventArgs) Handles dgvInventory.SelectionChanged
            If dgvInventory.SelectedRows.Count > 0 Then
                Dim row As DataGridViewRow = dgvInventory.SelectedRows(0)
                txtProductName.Text = If(row.Cells("ProductName").Value?.ToString(), "")
                Dim categoryValue As String = If(row.Cells("Category").Value?.ToString(), "")
                Dim categoryIndex As Integer = cmbCategory.Items.IndexOf(categoryValue)
                If categoryIndex >= 0 Then
                    cmbCategory.SelectedIndex = categoryIndex
                End If
                txtStock.Text = If(row.Cells("Stock").Value?.ToString(), "")
                txtPrice.Text = If(row.Cells("UnitPrice").Value?.ToString(), "")
            End If
        End Sub

        ' 一覧を更新
        Private Sub RefreshGrid(displayList As List(Of Product))
            dgvInventory.DataSource = Nothing
            dgvInventory.DataSource = displayList.Select(
                Function(p) New With {
                    p.Id,
                    p.ProductName,
                    p.Category,
                    p.Stock,
                    p.UnitPrice,
                    p.StockValue
                }).ToList()

            ' 列ヘッダーの表示名を設定
            If dgvInventory.Columns.Count > 0 Then
                dgvInventory.Columns("Id").HeaderText = "商品ID"
                dgvInventory.Columns("ProductName").HeaderText = "商品名"
                dgvInventory.Columns("Category").HeaderText = "カテゴリ"
                dgvInventory.Columns("Stock").HeaderText = "在庫数"
                dgvInventory.Columns("UnitPrice").HeaderText = "単価"
                dgvInventory.Columns("StockValue").HeaderText = "在庫金額"
            End If

            ' ★★★ バグ2: 在庫警告の表示処理がない ★★★
            ' 仕様 FR-008: 在庫数が5以下の商品は行を黄色で表示する
            ' 実装: 黄色表示の処理が完全に欠落している
            ' → 在庫が少ない商品が視覚的に分からない

            UpdateStatus(displayList)
        End Sub

        ' ステータスバー更新
        Private Sub UpdateStatus(displayList As List(Of Product))
            Dim count As Integer = displayList.Count
            Dim totalStockValue As Integer = displayList.Sum(Function(p) p.StockValue)

            lblStatus.Text = $"商品数: {count}件  在庫金額合計: {totalStockValue:N0}円"
        End Sub
    End Class
End Namespace

File: 01_CSharp_Level1/練習問題/問題2_検索表示バグ/CustomerManager/FormMain.cs
namespace CustomerManager
{
    public partial class FormMain : Form
    {
        private List<Customer> customers = new List<Customer>();
        private int nextId = 1;

        public FormMain()
        {
            InitializeComponent();
        }

        // フォーム読み込み時にサンプルデータを追加
        private void FormMain_Load(object sender, EventArgs e)
        {
            // サンプルデータ
            customers.Add(new Customer(nextId++, "田中太郎", "090-1234-5678", 35, "tanaka@example.com"));
            customers.Add(new Customer(nextId++, "鈴木花子", "080-9876-5432", 28, "suzuki@example.com"));
            customers.Add(new Customer(nextId++, "佐藤次郎", "070-1111-2222", 42, "sato@example.com"));

            RefreshGrid(customers);
        }

        // 追加ボタン
        private void btnAdd_Click(object sender, EventArgs e)
        {
            // バリデーション: 氏名必須チェック
            if (string.IsNullOrWhiteSpace(txtName.Text))
            {
                MessageBox.Show("氏名を入力してください", "入力エラー", MessageBoxButtons.OK, MessageBoxIcon.Warning);
                return;
            }

            // 年齢チェック: 整数かつ0-150の範囲
            if (!int.TryParse(txtAge.Text, out int age) || age < 0 || age > 150)
            {
                MessageBox.Show("年齢は0～150の整数で入力してください", "入力エラー", MessageBoxButtons.OK, MessageBoxIcon.Warning);
                return;
            }

            var customer = new Customer(nextId++, txtName.Text, txtPhone.Text, age, txtEmail.Text);
            customers.Add(customer);

            // 追加後、入力欄をクリア
            txtName.Text = "";
            txtPhone.Text = "";
            txtAge.Text = "";
            txtEmail.Text = "";

            RefreshGrid(customers);
        }

        // 編集ボタン
        private void btnEdit_Click(object sender, EventArgs e)
        {
            if (dgvCustomers.SelectedRows.Count == 0)
            {
                MessageBox.Show("顧客を選択してください", "警告", MessageBoxButtons.OK, MessageBoxIcon.Warning);
                return;
            }

            int selectedId = (int)dgvCustomers.SelectedRows[0].Cells["Id"].Value;
            var customer = customers.Find(c => c.Id == selectedId);

            if (customer != null)
            {
                customer.Name = txtName.Text;
                customer.Phone = txtPhone.Text;
                customer.Email = txtEmail.Text;

                if (int.TryParse(txtAge.Text, out int age))
                {
                    customer.Age = age;
                }

                RefreshGrid(customers);
            }
        }

        // 削除ボタン
        private void btnDelete_Click(object sender, EventArgs e)
        {
            if (dgvCustomers.SelectedRows.Count == 0)
            {
                MessageBox.Show("顧客を選択してください", "警告", MessageBoxButtons.OK, MessageBoxIcon.Warning);
                return;
            }

            int selectedId = (int)dgvCustomers.SelectedRows[0].Cells["Id"].Value;
            var customer = customers.Find(c => c.Id == selectedId);

            if (customer != null)
            {
                var result = MessageBox.Show(
                    $"{customer.Name}を削除しますか？",
                    "削除確認",
                    MessageBoxButtons.YesNo,
                    MessageBoxIcon.Question);

                if (result == DialogResult.Yes)
                {
                    customers.Remove(customer);
                    RefreshGrid(customers);
                }
            }
        }

        // 検索ボタン
        private void btnSearch_Click(object sender, EventArgs e)
        {
            string keyword = txtSearch.Text;

            if (string.IsNullOrWhiteSpace(keyword))
            {
                RefreshGrid(customers);
                return;
            }

            // ★★★ バグ1: 大文字・小文字を区別してしまう ★★★
            // 仕様 FR-004: 大文字・小文字を区別しない
            // 実装: StringComparison.OrdinalIgnoreCase が指定されていない
            // → "Tanaka" と "tanaka" が別扱いになる
            var results = customers.FindAll(c =>
                c.Name.Contains(keyword) ||
                c.Email.Contains(keyword));

            // ★★★ バグ2: 検索結果0件のメッセージがない ★★★
            // 仕様 FR-004: 検索結果が0件の場合「該当する顧客はありません」と表示
            // 実装: 0件チェックとメッセージ表示が抜けている
            // → ユーザーは結果がないことに気づきにくい

            RefreshGrid(results);
        }

        // 全件表示ボタン
        private void btnShowAll_Click(object sender, EventArgs e)
        {
            txtSearch.Text = "";
            RefreshGrid(customers);
        }

        // DataGridView選択変更時に入力欄に値をセット
        private void dgvCustomers_SelectionChanged(object sender, EventArgs e)
        {
            if (dgvCustomers.SelectedRows.Count > 0)
            {
                var row = dgvCustomers.SelectedRows[0];
                txtName.Text = row.Cells["Name"].Value?.ToString() ?? "";
                txtPhone.Text = row.Cells["Phone"].Value?.ToString() ?? "";
                txtAge.Text = row.Cells["Age"].Value?.ToString() ?? "";
                txtEmail.Text = row.Cells["Email"].Value?.ToString() ?? "";
            }
        }

        // 一覧を更新
        private void RefreshGrid(List<Customer> displayList)
        {
            dgvCustomers.DataSource = null;
            dgvCustomers.DataSource = displayList.Select(c => new
            {
                c.Id,
                c.Name,
                c.Phone,
                c.Age,
                c.Email
            }).ToList();

            UpdateStatus(displayList);
        }

        // ステータスバー更新
        private void UpdateStatus(List<Customer> displayList)
        {
            int count = displayList.Count;
            int totalAge = displayList.Sum(c => c.Age);

            // ★★★ バグ3: 整数除算で小数が切り捨てられる ★★★
            // 仕様 FR-006: 平均年齢は小数第1位まで表示
            // 実装: int同士の除算のため小数部分が切り捨てられる
            // → 例: 63/2=31 (本来は31.5)
            int avgAge = count > 0 ? totalAge / count : 0;

            lblStatus.Text = $"件数: {count}件  合計年齢: {totalAge}  平均年齢: {avgAge}";
        }
    }
}
